scipy/optimize/cython_optimize/__init__.py

```diff
@@ -0,0 +1,630 @@
+"""
```

Comment:
Good questions. This PR does a lot right now - review should probably focus on getting the API right before looking at everything else

Reply:
@rgommers is there an API that is preferred? I copied the callback signatures from [`quad`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.quad.html), and also I was following the advice of @person142 which was to have one signature that uses Python `tuple` and another that was pure C and uses `double, void*`. That's how I came up with the three signatures. Perhaps I should drop the array signature: `int, double*`, and just have two signatures, but the advantage to the array signature is that it's easier to use, b/c there's less code to write. To use the struct signature: `double, void*`, the user would have to make a `ctypedef`. EG: `zeros_array_examples.pyx` has 30 lines less than `zeros_struct_examples.pyx`

Reply:
>This __init__.py file is over 600 lines long and adds 15(?) functions. Much of it seems repetitive.
I suggest talking a look at the use of the keyword argument method=XXX used in scipy.optimize.root and (scipy.optimize.root_scalar in gh-8890) for a way to avoid creating many functions differing in only the method they call.

I've shortened the `__init__.py` file to about 160 lines. If I move the example to a tutorial then it will be only about 60 lines. Is that better? I've also removed the `newton/secant/halley` functions, so there are only the four pure C functions.

Thanks for your suggestion regarding #8890, but those functions are pure Python, not Cython, and in my opinion that PR is more about simplifying the API whereas this PR is about making the existing C root-finder functions accessible in Cython so they can "vectorized".

Do you think it would be preferred to remove two of the signatures, and just expose a single callback type `(double, void*)`? This should simplify the API significantly.

>Why the need for 3 variations of the functions taking struct, array and tuple?

This came from a suggestion from @person142 that there be two callback types, a simple one that uses a Python tuple, and another that uses a C struct, but perhaps I misunderstood, and only the C struct callback type is necessary. I added the array callback because I saw it in `scipy.integrate.quad` and it seemed easier to use than the C struct, because the user doesn't have to write a `ctypedef`.

Please let me know what your preference for callback type is. I'm happy to simplify to a single callback, as long as it is pure C and can release the GIL. My preference is for the C struct.